File: src/tokens/chunking.ts
/**
 * Text chunking utilities for context management.
 *
 * Split long texts into token-sized chunks for embeddings,
 * or truncate message histories to fit context windows.
 *
 * @module tokens/chunking
 */

import { estimateTokens } from './estimate.js';
import { estimateChatTokens, estimateMessageTokens } from './messages.js';
import type { ChunkOptions, FitOptions, Provider, TokenMessage } from './types.js';

/**
 * Split text into chunks of approximately maxTokens each.
 *
 * Attempts to split on natural boundaries (paragraphs, sentences)
 * rather than mid-word.
 *
 * @param text - Text to split
 * @param options - Chunking options
 * @returns Array of text chunks
 *
 * @example
 * ```typescript
 * import { splitText } from 'chatoyant/tokens';
 *
 * const chunks = splitText(longDocument, {
 *   maxTokens: 512,
 *   overlap: 50,
 * });
 * ```
 */
export function splitText(text: string, options: ChunkOptions): string[] {
  const { maxTokens, overlap = 0, separator } = options;

  if (!text) return [];

  const totalTokens = estimateTokens(text);
  if (totalTokens <= maxTokens) {
    return [text];
  }

  // Determine split points
  const splitPattern =
    separator instanceof RegExp
      ? separator
      : separator
        ? new RegExp(separator, 'g')
        : /\n\n+|\n|(?<=[.!?])\s+/g;

  // Split into segments
  const segments = text.split(splitPattern).filter((s) => s.trim());

  const chunks: string[] = [];
  let currentChunk = '';
  let currentTokens = 0;

  for (const segment of segments) {
    const segmentTokens = estimateTokens(segment);

    // If single segment exceeds max, split it further
    if (segmentTokens > maxTokens) {
      // Flush current chunk
      if (currentChunk) {
        chunks.push(currentChunk.trim());
        currentChunk = '';
        currentTokens = 0;
      }

      // Split large segment by words
      const words = segment.split(/\s+/);
      let wordChunk = '';
      let wordTokens = 0;

      for (const word of words) {
        const wordTokenCount = estimateTokens(`${word} `);
        if (wordTokens + wordTokenCount > maxTokens && wordChunk) {
          chunks.push(wordChunk.trim());
          wordChunk = overlap > 0 ? getOverlapText(wordChunk, overlap) : '';
          wordTokens = estimateTokens(wordChunk);
        }
        wordChunk += `${word} `;
        wordTokens += wordTokenCount;
      }

      if (wordChunk.trim()) {
        currentChunk = wordChunk;
        currentTokens = wordTokens;
      }
      continue;
    }

    // Check if adding this segment exceeds max
    const newTokens = currentTokens + segmentTokens + 1; // +1 for separator
    if (newTokens > maxTokens && currentChunk) {
      chunks.push(currentChunk.trim());

      // Start new chunk with overlap from previous
      if (overlap > 0) {
        currentChunk = `${getOverlapText(currentChunk, overlap)} ${segment}`;
        currentTokens = estimateTokens(currentChunk);
      } else {
        currentChunk = segment;
        currentTokens = segmentTokens;
      }
    } else {
      currentChunk += (currentChunk ? ' ' : '') + segment;
      currentTokens = newTokens;
    }
  }

  // Add remaining chunk
  if (currentChunk.trim()) {
    chunks.push(currentChunk.trim());
  }

  return chunks;
}

/**
 * Get overlap text from the end of a chunk.
 */
function getOverlapText(text: string, overlapTokens: number): string {
  const words = text.trim().split(/\s+/);
  let result = '';
  let tokens = 0;

  // Work backwards to get overlap tokens worth of text
  for (let i = words.length - 1; i >= 0 && tokens < overlapTokens; i--) {
    const wordTokens = estimateTokens(words[i]);
    result = words[i] + (result ? ` ${result}` : '');
    tokens += wordTokens;
  }

  return result;
}

/**
 * Truncate messages to fit within a token budget.
 *
 * Removes oldest messages first (except system message).
 *
 * @param messages - Messages to truncate
 * @param options - Fit options
 * @returns Truncated messages that fit within budget
 *
 * @example
 * ```typescript
 * import { fitMessages } from 'chatoyant/tokens';
 *
 * const fitted = fitMessages(conversationHistory, {
 *   maxTokens: 4000,
 *   reserveForResponse: 1000,
 * });
 * ```
 */
export function fitMessages(messages: TokenMessage[], options: FitOptions): TokenMessage[] {
  const { maxTokens, reserveForResponse = 0, provider = 'openai' } = options;

  const budget = maxTokens - reserveForResponse;

  if (budget <= 0) return [];
  if (messages.length === 0) return [];

  // Check if already fits
  const totalTokens = estimateChatTokens(messages, provider);
  if (totalTokens <= budget) {
    return messages;
  }

  // Separate system message (if any) from rest
  const systemMessage = messages[0]?.role === 'system' ? messages[0] : null;
  const otherMessages = systemMessage ? messages.slice(1) : messages;

  const systemTokens = systemMessage ? estimateMessageTokens(systemMessage, provider) : 0;
  const remainingBudget = budget - systemTokens;

  if (remainingBudget <= 0) {
    // Can't even fit system message
    return systemMessage ? [systemMessage] : [];
  }

  // Add messages from the end (most recent first)
  const result: TokenMessage[] = [];
  let usedTokens = 0;

  for (let i = otherMessages.length - 1; i >= 0; i--) {
    const msg = otherMessages[i];
    const msgTokens = estimateMessageTokens(msg, provider);

    if (usedTokens + msgTokens <= remainingBudget) {
      result.unshift(msg);
      usedTokens += msgTokens;
    } else {
      // Can't fit more messages
      break;
    }
  }

  // Add system message back at the start
  if (systemMessage) {
    result.unshift(systemMessage);
  }

  return result;
}

/**
 * Truncate a single message content to fit token limit.
 *
 * @param content - Message content to truncate
 * @param maxTokens - Maximum tokens allowed
 * @param ellipsis - String to append when truncated (default: '...')
 * @returns Truncated content
 */
export function truncateContent(
  content: string,
  maxTokens: number,
  ellipsis: string = '...',
): string {
  if (!content) return content;

  const tokens = estimateTokens(content);
  if (tokens <= maxTokens) return content;

  // Estimate characters to keep
  const ratio = maxTokens / tokens;
  const targetChars = Math.floor(content.length * ratio) - ellipsis.length;

  if (targetChars <= 0) return ellipsis;

  // Try to break at word boundary
  let truncated = content.slice(0, targetChars);
  const lastSpace = truncated.lastIndexOf(' ');

  if (lastSpace > targetChars * 0.8) {
    truncated = truncated.slice(0, lastSpace);
  }

  return truncated.trim() + ellipsis;
}

/**
 * Split messages at a token boundary for pagination.
 *
 * @param messages - Messages to split
 * @param tokensPerPage - Tokens per page
 * @param provider - Provider for overhead calculation
 * @returns Array of message page arrays
 */
export function paginateMessages(
  messages: TokenMessage[],
  tokensPerPage: number,
  provider: Provider = 'openai',
): TokenMessage[][] {
  if (messages.length === 0) return [];

  const pages: TokenMessage[][] = [];
  let currentPage: TokenMessage[] = [];
  let currentTokens = 0;

  for (const msg of messages) {
    const msgTokens = estimateMessageTokens(msg, provider);

    if (currentTokens + msgTokens > tokensPerPage && currentPage.length > 0) {
      pages.push(currentPage);
      currentPage = [];
      currentTokens = 0;
    }

    currentPage.push(msg);
    currentTokens += msgTokens;
  }

  if (currentPage.length > 0) {
    pages.push(currentPage);
  }

  return pages;
}

/**
 * Calculate how many chunks a text will produce.
 *
 * @param text - Text to analyze
 * @param chunkSize - Target tokens per chunk
 * @returns Estimated number of chunks
 */
export function estimateChunkCount(text: string, chunkSize: number): number {
  if (!text || chunkSize <= 0) return 0;
  const tokens = estimateTokens(text);
  return Math.ceil(tokens / chunkSize);
}

File: lib/tokens/chunking.ml
type options = { max_tokens : int; overlap : int; separator : string option }

let trim = String.trim

let split_on_substring ~separator text =
  if separator = "" then [ text ]
  else
    let separator_len = String.length separator in
    let text_len = String.length text in
    let rec find_from index =
      if index + separator_len > text_len then None
      else if String.sub text index separator_len = separator then Some index
      else find_from (index + 1)
    in
    let rec loop start acc =
      match find_from start with
      | None -> List.rev (String.sub text start (text_len - start) :: acc)
      | Some index ->
          let part = String.sub text start (index - start) in
          loop (index + separator_len) (part :: acc)
    in
    loop 0 []

let words text =
  text |> String.split_on_char ' ' |> List.map trim |> List.filter (( <> ) "")

let default_segments text =
  text |> String.split_on_char '\n' |> List.concat_map words

let segments ?separator text =
  match separator with
  | Some separator ->
      split_on_substring ~separator text
      |> List.map trim
      |> List.filter (( <> ) "")
  | None -> default_segments text

let overlap_text text overlap_tokens =
  if overlap_tokens <= 0 then ""
  else
    let words = words text in
    let rec loop tokens acc = function
      | [] -> String.concat " " acc
      | word :: rest ->
          if tokens >= overlap_tokens then String.concat " " acc
          else
            let word_tokens = Token_estimate.estimate word in
            loop (tokens + word_tokens) (word :: acc) rest
    in
    loop 0 [] (List.rev words)

let split_large_segment ~max_tokens ~overlap segment =
  let rec loop current current_tokens chunks = function
    | [] ->
        if trim current = "" then List.rev chunks
        else List.rev (trim current :: chunks)
    | word :: rest ->
        let token_count = Token_estimate.estimate (word ^ " ") in
        if current <> "" && current_tokens + token_count > max_tokens then
          let overlap = overlap_text current overlap in
          let next = if overlap = "" then word else overlap ^ " " ^ word in
          loop next (Token_estimate.estimate next) (trim current :: chunks) rest
        else
          let next = if current = "" then word else current ^ " " ^ word in
          loop next (current_tokens + token_count) chunks rest
  in
  loop "" 0 [] (words segment)

let split_text ?(overlap = 0) ?separator ~max_tokens text =
  if text = "" then []
  else if max_tokens <= 0 then [ text ]
  else if Token_estimate.estimate text <= max_tokens then [ text ]
  else
    let rec loop current current_tokens chunks = function
      | [] ->
          if trim current = "" then List.rev chunks
          else List.rev (trim current :: chunks)
      | segment :: rest ->
          let segment_tokens = Token_estimate.estimate segment in
          if segment_tokens > max_tokens then
            let flushed =
              if trim current = "" then chunks else trim current :: chunks
            in
            let large_chunks =
              split_large_segment ~max_tokens ~overlap segment
            in
            loop "" 0 (List.rev_append large_chunks flushed) rest
          else
            let added_tokens =
              current_tokens + segment_tokens + if current = "" then 0 else 1
            in
            if current <> "" && added_tokens > max_tokens then
              let overlap_text = overlap_text current overlap in
              let next =
                if overlap_text = "" then segment
                else overlap_text ^ " " ^ segment
              in
              loop next
                (Token_estimate.estimate next)
                (trim current :: chunks) rest
            else
              let next =
                if current = "" then segment else current ^ " " ^ segment
              in
              loop next added_tokens chunks rest
    in
    loop "" 0 [] (segments ?separator text)

let truncate_content ?(ellipsis = "...") ~max_tokens content =
  if content = "" then content
  else
    let tokens = Token_estimate.estimate content in
    if tokens <= max_tokens then content
    else
      let target_chars =
        int_of_float
          (floor
             (Float.of_int (String.length content)
             *. (Float.of_int (max 0 max_tokens) /. Float.of_int (max 1 tokens))
             ))
        - String.length ellipsis
      in
      if target_chars <= 0 then ellipsis
      else
        let truncated =
          String.sub content 0 (min target_chars (String.length content))
        in
        let last_space =
          try Some (String.rindex truncated ' ') with Not_found -> None
        in
        let truncated =
          match last_space with
          | Some index
            when Float.of_int index > Float.of_int target_chars *. 0.8 ->
              String.sub truncated 0 index
          | _ -> truncated
        in
        trim truncated ^ ellipsis

let fit_messages ?(provider = Message_budget.Openai) ~max_tokens
    ?(reserve_for_response = 0) messages =
  Message_budget.fit ~provider ~max_tokens ~reserve_for_response messages

let paginate_messages ?(provider = Message_budget.Openai) ~tokens_per_page
    messages =
  let rec loop current current_tokens pages = function
    | [] ->
        if current = [] then List.rev pages
        else List.rev (List.rev current :: pages)
    | message :: rest ->
        let message_tokens =
          Message_budget.estimate_message ~provider message
        in
        if current <> [] && current_tokens + message_tokens > tokens_per_page
        then loop [ message ] message_tokens (List.rev current :: pages) rest
        else
          loop (message :: current) (current_tokens + message_tokens) pages rest
  in
  loop [] 0 [] messages

let estimate_chunk_count ~chunk_size text =
  if text = "" || chunk_size <= 0 then 0
  else
    int_of_float
      (ceil
         (Float.of_int (Token_estimate.estimate text) /. Float.of_int chunk_size))
